server/utils/validateStateCode.js

```
@@ -0,0 +1,57 @@
+/*
+ * Recidiviz - a data platform for criminal justice reform
+ * Copyright (C) 2021 Recidiviz, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ * =============================================================================
+ *
+ */
+
+const { pathToRegexp } = require("path-to-regexp");
+
+const { getAppMetadata } = require("./getAppMetadata");
+
+const validateStateCode = (exemptRoutes) => {
+  return (req, res, next) => {
+    const matchingRoutes = exemptRoutes.filter(
+      (route) => pathToRegexp(route).exec(req.originalUrl) !== null
+    );
+
+    if (matchingRoutes.length > 0) {
+      next();
+      return;
+    }
+
+    const reqStateCode = req.params.stateCode.toLowerCase();
+
+    // only filter on state codes (us_xx)
+    if (/us_[a-z][a-z]/.exec(reqStateCode) === null) {
```

Why is it that there is separate treatment for "exempted routes" and routes without stateCode in the path? Why can't the latter be explicitly listed in the former? The upside to that is it should make it so that if someone in the future mistakenly adds a route without stateCode when they meant to include stateCode, it would mean that instead of just letting requests silently next their way through, it would fail at dev time, because presumably we'd have some sort of test that ensures all routes either have stateCode or are exempted.

---

@jessex I agree. The only route as this is written that is not "unprotected" but does not include stateCode is `/api/generateFileLink`. I think it makes sense to add that to the exempted list and add the `checkJwt` back into it.

Maybe the exempted list can be renamed to `routesWithoutStateCode` and we add the `checkJwt` back into each route individually to be very explicit. 

Since @Catacola is OOO today, I'll push up a commit to see what you think.